### PR TITLE
chore: prep react-query v5 by using object syntax

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Review by KATT — COMMENTED** [review]
LGTM! This might mean we want to bump the peer dep of tanstack query to a higher 4.x series if they got added in prep for v5

**Comment by juliusmarminge** [comment]
> LGTM! This might mean we want to bump the peer dep of tanstack query to a higher 4.x series if they got added in prep for v5

I don't think this syntax is new so should be supported in the current peerDep 4.3. @TkDodo can probably confirm.

There's some stuff left with the utils (which is why tests are failing) - will get to those next week. Prepping for exam on Monday.

**Comment by KATT** [comment]
Good luck with the exam!!

**Comment by TkDodo** [comment]
> I don't think this syntax is new so should be supported in the current peerDep 4.3. @TkDodo can probably confirm.

The object syntax also exists in v4. v5 will just make it the _only_ valid syntax.

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 1]
```diff
@@ -78,65 +79,66 @@ export function createRootHooks<
...
-              return queryClient.fetchQuery(
-                getArrayQueryKey(pathAndInput, 'query'),
-                () =>
+              return queryClient.fetchQuery({
+                queryKey: getArrayQueryKey(pathAndInput, 'query'),
+                queryFn: () =>
                   (client as any).query(...getClientArgs(pathAndInput, opts)),
-                opts,
-              );
+                ...opts,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:86]: Maybe move this to the first arguments so the `queryKey`/Fn can't be overriden

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 2]
```diff
@@ -78,65 +79,66 @@ export function createRootHooks<
...
                   const [path, input] = pathAndInput;
-                  const actualInput = { ...(input as any), cursor: pageParam };
+                  const actualInput = { ...input, cursor: pageParam };
                   return (client as any).query(
                     ...getClientArgs([path, actualInput], opts),
                   );
                 },
-                opts,
-              );
+                ...opts,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:102]: Move first?

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 3]
```diff
@@ -78,65 +79,66 @@ export function createRootHooks<
...
-            (...args: any[]) => {
-              const [queryKey, ...rest] = args;
-
+            (queryKey, filters, options) => {
               return queryClient.invalidateQueries(
-                getArrayQueryKey(queryKey, 'any'),
-                ...rest,
+                {
+                  queryKey: getArrayQueryKey(queryKey as any, 'any'),
+                  ...filters,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:139]: First?

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 4]
```diff
@@ -145,29 +147,29 @@ export function createRootHooks<
             (...args: any[]) => {
               const [queryKey, ...rest] = args;
 
-              return queryClient.resetQueries(
-                getArrayQueryKey(queryKey, 'any'),
+              return queryClient.resetQueries({
+                queryKey: getArrayQueryKey(queryKey, 'any'),
                 ...rest,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:152]: First?

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 5]
```diff
@@ -145,29 +147,29 @@ export function createRootHooks<
...
           ),
           refetchQueries: useCallback(
             (...args: any[]) => {
               const [queryKey, ...rest] = args;
 
-              return queryClient.refetchQueries(
-                getArrayQueryKey(queryKey, 'any'),
+              return queryClient.refetchQueries({
+                queryKey: getArrayQueryKey(queryKey, 'any'),
                 ...rest,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:163]: same

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 6]
```diff
@@ -286,8 +288,10 @@ export function createRootHooks<
           ...getClientArgs(pathAndInput, actualOpts),
         );
       },
-      { context: ReactQueryContext, ...ssrOpts },
-    ) as UseTRPCQueryResult<unknown, TError>;
+      context: ReactQueryContext,
+      ...((ssrOpts ?? {}) as UseQueryOptions), // <-- why is this cast needed?
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:292]: Remove it if not

**Review by KATT — DISMISSED** [review]

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx (hunk 3 above)
Comment (KATT) [reply packages/react-query/src/shared/hooks/createHooksInternal.tsx:138]: 👀

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 7]
```diff
@@ -269,9 +276,10 @@ export function createRootHooks<
     // request option should take priority over global
     const shouldAbortOnUnmount = opts?.trpc?.abortOnUnmount ?? abortOnUnmount;
 
-    const hook = __useQuery(
-      getArrayQueryKey(pathAndInput, 'query') as any,
-      (queryFunctionContext) => {
+    const hook = __useQuery({
+      ...ssrOpts,
+      queryKey: getArrayQueryKey(pathAndInput, 'query') as any,
+      queryFn: (queryFunctionContext) => {
```
Comment (juliusmarminge) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:282]: types complain here without the any - no idea why... either type the opts or the querykey as any 🤷🏼‍♂️

Comment (KATT) [reply packages/react-query/src/shared/hooks/createHooksInternal.tsx:282]: meh 🤷  it's okay sometimes, feel free to leave a `// FIXME` comment if you want

**Review by KATT — APPROVED** [review]

**Pull request merged**